File: resources/code/2025-10-21-logic-and-proof/example1-3.hs
data F = F F
type T = ()

example1 :: p -> (q -> p)
example1 a = \b -> a

example2 :: (p -> r) -> (q -> r) -> (Either p q -> r)
example2 f g (Left a) = f a
example2 f g (Right b) = g b

example3 :: (p -> q -> r) -> ((p, q) -> r)
example3 f (a, b) = f a b

main :: IO()
main = putStrLn "Hello, World"

File: resources/code/2025-10-21-logic-and-proof/example1-3.ts
type F = never;
class T {}

const example1: <P, Q>(a: P) => ((b: Q) => P)
    = <P, Q>(a: P) => {
        return (_: Q) => a;
    };

// const example2: <P, Q, R>(f: (_: P) => R) => (g: (_: Q) => R) => (c: P | Q) => R
//     = ...;  // Unable to write in TypeScript

const example3: <P, Q, R>(f: (_: P) => (_: Q) => R) => ((a: [P, Q]) => R)
    = <P, Q, R>(f: (_: P) => (_: Q) => R) => {
        return (a: [P, Q]) => {
            return f(a[0])(a[1]);
        };
    };